The Big Merge

brianwinterpixel opened the pull request:
Merged all my work from earlier this year that hadn't been downstreamed

Too many fixes and changes to explain it all, but the biggest _addition_ was world querying functions. From simple ray casting to continuous full-body casting.


(i need this to make shields better I swear)

briansemrau committed:
Merge latest 4.0 to 3.x (#63)

* Port 4.0 impl to 3.x (all changes after commit 4418875ba302b7dedd89b6e0d6ccdb10fd18e554)

Not all features from 4.0 have been tested to work after porting to 3.x

Co-authored-by: Jordan Schidlowsky <jordan@winterpixel.com>

briansemrau committed:
Restore icons

briansemrau committed:
Fix incorrect damping setting

briansemrau committed:
Fix error when using flag tools=no

briansemrau committed:
Fix crash when body is deleted while being contact monitored (3.x) (#64)

* Fix areas not emitting enter/exit signals

* Fix null crash when monitored bodies are deleted.

briansemrau committed:
Fix area/body_fixture_entered/exited missing param

briansemrau committed:
Refactor world step, add virtual script function _world_step (#68)

briansemrau committed:
Various bugfixes and small changes (#69)

* Refactor world step, add virtual script function _world_step

* Fix null crash when objects destroyed

* Fixed incomplete method bindings

* Fix world raycast callback

* Fix errors when active contacts > max reported contacts

* Update box2d_physics_body.cpp

* Remove all dynamic casts for web builds

* fixing paths

* Expose b2world velocity/position iteration params

Co-Authored-By: Jordan Schidlowsky <323868+jordo@users.noreply.github.com>

* Allow larger shape sizes

Co-authored-by: Jordan Schidlowsky <jordanschidlowsky@gmail.com>
Co-authored-by: Jordan Schidlowsky <323868+jordo@users.noreply.github.com>

briansemrau committed:
Fix mishandled case in contact buffering

Fixes #71

briansemrau committed:
Fix Box2DArea not tracking parent transform correctly

briansemrau committed:
Fix contact monitor issues

- Contact point swapping was incorrectly handled before, but now is fixed.
- Fixed contact initial data not being written
- Contact impulse units are now correct
- Added more error checks for contact data getters

brianwinterpixel committed:
Merge remote-tracking branch 'briansemrau/3.x' into game-tankkings

jordo commented on scene/2d/box2d_collision_object.h:
```diff
@@ -0,0 +1,116 @@
+	int get_colliding_bodies_fast(Array p_array) const;
+
+	Box2DCollisionObject();
+	~Box2DCollisionObject();
```
good practice to always make destructors virtual.

jordo replied:
probably need everywhere.

brianwinterpixel replied:
I'll need to learn a bit more about the risks of nonvirtual destructors.
Created a new issue for this.

jordo commented on scene/2d/box2d_physics_body.cpp:
```diff
@@ -223,68 +340,48 @@ void Box2DPhysicsBody::_notification(int p_what) {
 				filtering_me[i]->filtered.erase(this);
 			}
-			
-			destroy_b2Body();
```
When does the destroy happen now?

brianwinterpixel replied:
~~see https://github.com/winterpixelgames/godot_box2d/commit/c017c72adf122d66b65c6f5e4dd65ffd6dd4bea4~~ nvm not important

most things missing in box2d_physics_body.cpp are likely found in box2d_collision_object.cpp

jordo commented on scene/2d/box2d_physics_body.cpp:
```diff
@@ -223,68 +340,48 @@ void Box2DPhysicsBody::_notification(int p_what) {
-
 			set_process_internal(false);
-			if (world_node) {
-				destroy_b2Body();
```
moved or corrected somewhere?

brianwinterpixel replied:
should be in Box2DCollisionObject now

jordo commented on scene/2d/box2d_physics_body.h:
```diff
@@ -36,69 +40,108 @@ class Box2DPhysicsBody : public Node2D {
+
+	// Used for move_and_slide/_with_snap
+	std::vector<KinematicCollision> kinematic_colliders;
+	Vector<Ref<Box2DKinematicCollision>> kinematic_colliders_refcache;
```
i wouldn't use godot Vector anywhere.  It's a realloc on every push.  LocalVector is better if need a godot data structure.

jordo commented on scene/2d/box2d_world.cpp:
```diff
@@ -15,6 +18,189 @@
+	ret.resize(parameters.exclude.size());
+	int idx = 0;
+	for (Set<const Box2DPhysicsBody *>::Element *E = parameters.exclude.front(); E; E = E->next()) {
+		ret[idx] = int64_t(E->get()->get_instance_id());
```
increment idx here?

brianwinterpixel replied:
!

jordo commented on scene/2d/box2d_world.cpp:
```diff
@@ -268,97 +486,71 @@ void Box2DWorld::PreSolve(b2Contact *contact, const b2Manifold *oldManifold) {
 	if (buffer_manifold) {
 		// Only handle the first PreSolve for this contact this step (don't overwrite initial impact_velocity, world_pos)
-		for (int i = 0; i < buffer_manifold->count; ++i) {
+		for (int i = 0; i < b2_maxManifoldPoints; ++i) {
```
whats?   b2_maxManifoldPoints

jordo replied:
why not count still?

brianwinterpixel replied:
I believe it's because the b2 contact can have the idx=0 manifold point be invalid while idx=1 is a valid point. Not sure the details, but the changes to the contact monitor are important bug fixes.

brianwinterpixel commented:
I forgot to remove one of the signal callback mechanisms. Both jordo's and my own are implemented so they definitely won't work as expected if we tried to use them.

jordo reviewed (commented):
Big merge, so I can't really comment on any of the logic, but I added a few things.

I tested and game still works, but there's a contact regression.  See the glitches in video below, this happens when client resets/resims through a contact.  There's a glitch when the tank hits a surface.  This can happen when the client gets unlucky, and gets a state update to a past state with no contact, replayed through the the current frame with the existing contact.  

Biggest risks would be not really knowing what the performance characteristics of somethings are now, and any memory leaks, etc.  

https://user-images.githubusercontent.com/323868/145332048-05c2aeca-41ea-46cb-9244-5f89db6860fa.mov

jordo commented:
https://user-images.githubusercontent.com/323868/145332660-d12b8537-f3d2-444c-90bc-aae5e7fbcc1a.mov

Another example, more pronounced by doing 2 things in the TestClientServerGame scene.

Set the server broadcast rate to be less often:
<img width="1512" alt="Screen Shot 2021-12-08 at 10 06 39 PM" src="https://user-images.githubusercontent.com/323868/145332724-377eba57-b984-413a-8eda-e3f5be7979f0.png">

And set the Client simulated lag to be higher:
<img width="1504" alt="Screen Shot 2021-12-08 at 10 06 59 PM" src="https://user-images.githubusercontent.com/323868/145332748-e5444825-ab6b-467c-8119-f8b61da174b1.png">

jordo commented:
Oh wait, the regression happens when the client is reset to the exact frame on which the contact is created iirc.

brianwinterpixel commented:
I'm surprised that the contact stuff regressed, as all the contact changes in this aren't used in the game code afaik

jordo commented:
i'll make sure it didn't regress by something else...  Lemme try on current branch and take two videos to make sure.

brianwinterpixel commented:
@jordo Looks like you're using an old version, given the mine bot and the tank skin in the lobby

jordo commented:
Aren't those still in TestClientServerGame.tscn still?

brianwinterpixel commented:
not as of https://github.com/winterpixelgames/tankkings/commit/a8fef485931b4054fd648654932732388b10112e#diff-099849744652fad82f94c756406b7f0f729eba8c18d2e4975f843730408de5d7R53

jordo commented:
yup, hazards of switching machines constantly.   I'll verify if that is new or still in previous version.

jordo commented:
Boo.  It's regressed, but it must have been from something previously, so this PR should be good.  This was pretty hard to track down before.

https://user-images.githubusercontent.com/323868/145336032-23fdb07f-80f9-40e8-80c1-4a9c1b73d551.mov

jordo commented:
Looks good to merge.

brianwinterpixel commented:
Before merging I'll do a full pass through it again tomorrow

brianwinterpixel committed:
Move enabled_state_changed calc to collision object

Also fixed the `prev_enabled_state` flag not being reset, taking up >30% of the profiled time inside `Box2DWorld::step` from just `emit_signal` calls

brianwinterpixel committed:
Add sdf shape to the hell structure

not sure if this is useful

brianwinterpixel committed:
Reduce code duplication for templated contact queue

brianwinterpixel committed:
Remove GodotSignalCaller collision callback handler

brianwinterpixel committed:
Fix issues with kinematic collision implementation

brianwinterpixel commented on scene/2d/box2d_fixtures.h:
```diff
@@ -59,12 +60,16 @@ class Box2DFixture : public Node2D {
 
 	virtual String get_configuration_warning() const override;
 
+	Box2DCollisionObject *get_owner() const { return owner_node; }
```
redundant method. should be removed at some point

brianwinterpixel committed:
Fix dumb default mask value

brianwinterpixel committed:
Stupid typo that broke intersect_shape
